### PR TITLE
refactor: simplify WordWrapper implementation

### PR DESCRIPTION
**Pull request by tranzystorekk**
The `WordWrapper` implementation is a large and almost unreadable chunk of code, so I decided to do a ground-up cleaning.

Highlights and notes in no particular order:

- Before, due to extensive nesting and poor communication the control flow was foggy and mentally exhausting. Now the code is more clearly separated into a) emitting any currently cached lines (`self.wrapped_lines`) and b) processing the input iterator for any input to be wrapped and cached for step `a`
- I tried to invert and compact any larger if blocks, using let-else and labeled blocks. These features came in Rust 1.65 so they shouldn't affect our MSRV
- I attempted to rename some variables to more clearly signal intent, e.g. `unfinished_word` -> `pending_word`, `has_seen_non_whitespace` -> `non_whitespace_previous`
- Reduced the number of comments, in the old code a lot of them only contributed to the general noise or redundancy
- Some miscellaneous changes, like using `u32::saturating_sub` or `VecDeque::drain`

### CODE REVIEW

**Review by orhun — APPROVED** [review]
This guy doing the god's work over here.

If the CI is green, I hit approve on my screen.

File: src/widgets/reflow.rs [hunk 1]
```diff
@@ -72,155 +72,142 @@ where
...
                         }
-                        current_line.append(&mut unfinished_word);
-                    }
-                    if !current_line.is_empty() {
-                        wrapped_lines.push(current_line);
+
+                        whitespace_width -= width;
+                        remaining_width -= width;
+                        pending_whitespace.pop_front();
                     }
```
Comment (joshka) [inline src/widgets/reflow.rs:170]: I tried to change this so that it's just calling pop_front() instead of both front and pop_front. I'm not 100% sure why this isn't the equivalent of:
```rust
// remove whitespace up to the end of line
while let Some(grapheme) = pending_whitespace.pop_front() {
    let width = grapheme.symbol.width() as u16;
    whitespace_width -= width;
    if width > remaining_width {
        break;
    }
    remaining_width -= width;
}
```
This doesn't seem to append the first line of spaces in the indent test, but I'm not sure why.

File: src/widgets/reflow.rs [hunk 2]
```diff
@@ -72,155 +72,142 @@ where
...
+                    wrapped_lines.push(vec![]);
+                } else if !self.trim || !current_line.is_empty() {
+                    current_line.extend(pending_whitespace);
+                } else {
+                    // TODO: explain why this else branch is ok
+                    // See clippy::else_if_without_else
+                }
+
+                current_line.append(&mut pending_word);
+            }
```
Comment (joshka) [inline src/widgets/reflow.rs:202]: This is also a fairly dense if statement - is there a good way to simplify this logic?

**Review by joshka — COMMENTED** [review]
This is definitely incrementally better. It does however have a few downsides.

It does increase the length of this method a bit. That can be mitigated by thinking about what each block is doing, which variables are scoped to each block, and extracting appropriate functions that highlight and name that scope of changes.

In an of itself, method size is one of those controversial topics. Here I'd suggest finding ways to lower it. In reading this, I found myself repeatedly scrolling up to work out which block I was in so I could reason about the effects of lines of code.

Removing some of the comments made it subjectively more difficult to understand why certain checks are necessary - how they are work is the job of the code, why is the job of the comments.

Thanks for taking a stab at this. I'd merge this PR as it stands, but would like to hear your thoughts on the review comments first.

File: src/widgets/reflow.rs [hunk 3]
```diff
@@ -72,155 +72,142 @@ where
...
-                    let mut has_seen_non_whitespace = false;
-                    for StyledGrapheme { symbol, style } in line_symbols {
-                        let symbol_whitespace = symbol == ZWSP
-                            || (symbol.chars().all(&char::is_whitespace) && symbol != NBSP);
-                        let symbol_width = symbol.width() as u16;
-                        // Ignore characters wider than the total max width
-                        if symbol_width > self.max_line_width {
-                            continue;
-                        }
+            self.wrapped_lines = None;
```
[the remark this answers is not in the document]
Comment (tranzystorekk) [reply src/widgets/reflow.rs:98]: Yeah, I was overthinking that code a bit - my defensive programming instinct said any non-cleaned-up state like this might end up biting us in unexpected ways, but I agree that this just muddies the code in the end.

File: src/widgets/reflow.rs [hunk 4]
```diff
@@ -72,155 +72,142 @@ where
...
-                            current_line.append(&mut unfinished_word);
-                            current_line_width += word_width;
-
-                            // Clear whitespace buffer
-                            unfinished_whitespaces.clear();
-                            whitespace_width = 0;
-                            word_width = 0;
-                        }
+            // otherwise, process pending wrapped lines from input
+            let (line_symbols, line_alignment) = self.input_lines.next()?;
```
[the remark this answers is not in the document]
Comment (tranzystorekk) [reply src/widgets/reflow.rs:101]: We could "uncollapse" this into a let-else to make it more prominent (much less likely to miss if it explicitly says `return None;`)

[the remark this answers is not in the document]
Comment (joshka) [reply src/widgets/reflow.rs:101]: I'd preferentially look for ways where the size of the method makes this much more obvious. Right now this is a single character lost in a 150 line method. Increase the signal to noise ratio by reducing the total size of the method to something that fits on in something more like 20 lines of code and this is likely ok. But until then let-else-return is better.

File: src/widgets/reflow.rs [hunk 5]
```diff
@@ -72,155 +72,142 @@ where
...
+                    .map(|grapheme| grapheme.symbol.width() as u16)
+                    .sum();
+
+                self.current_line = line;
+                return Some(WrappedLine {
+                    line: &self.current_line,
+                    width: line_width,
+                    alignment: self.current_alignment,
+                });
             }
```
[the remark this answers is not in the document]
Comment (tranzystorekk) [reply src/widgets/reflow.rs:96]: Unfortunately I can only extract the first half of the short-circuit logic, because the borrow checker bends under the weight of borrowing other parts of `Self`, like `self.current_alignment`.

I still see this as a net gain though - the let-elses + `'emit` are too clunky for my taste after all.

File: src/widgets/reflow.rs [hunk 6]
```diff
@@ -72,155 +72,142 @@ where
...
-                            }
-                        }
+            let mut wrapped_lines = vec![];
+            let mut current_line = vec![];
+            let mut current_line_width = 0;
+            let mut pending_word = vec![];
+            let mut word_width = 0;
+            let mut pending_whitespace: VecDeque<StyledGrapheme> = VecDeque::new();
+            let mut whitespace_width = 0;
+            let mut non_whitespace_previous = false;
```
[the remark this answers is not in the document]
Comment (tranzystorekk) [reply src/widgets/reflow.rs:111]: I'd call this whole portion of code sth like `fn process_input(&mut self, line_symbols: impl IntoIterator<Item = StyledGrapheme>)` and go from there incrementally

File: src/widgets/reflow.rs [hunk 7]
```diff
@@ -72,155 +72,142 @@ where
...
 
-                        // Append symbol to unfinished, partially processed word
-                        if symbol_whitespace {
-                            whitespace_width += symbol_width;
-                            unfinished_whitespaces.push_back(StyledGrapheme { symbol, style });
-                        } else {
-                            word_width += symbol_width;
-                            unfinished_word.push(StyledGrapheme { symbol, style });
-                        }
+            for StyledGrapheme { symbol, style } in line_symbols {
```
[the remark this answers is not in the document]
Comment (tranzystorekk) [reply src/widgets/reflow.rs:113]: Sounds like this is the same point as your comment above?

File: src/widgets/reflow.rs [hunk 8]
```diff
@@ -72,155 +72,142 @@ where
...
+                        && current_line.is_empty()
+                        && self.trim
+                    || word_width + whitespace_width + symbol_width > self.max_line_width
+                        && current_line.is_empty()
+                        && !self.trim
+                {
+                    if !current_line.is_empty() || !self.trim {
+                        current_line.extend(pending_whitespace.drain(..));
+                        current_line_width += whitespace_width;
                     }
```
[the remark this answers is not in the document]
Comment (tranzystorekk) [reply src/widgets/reflow.rs:138]: Hm, seems like it will also trigger in the last statement clause (the one with `!self.trim()`, but I definitely am having a hard time following the overall control flow, so I'm not even sure if it's possible...

File: src/widgets/reflow.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (tranzystorekk) [reply src/widgets/reflow.rs:111]: My first thought when seeing the original name `has_seen_non_whitespace` was to fix how this flag is marked below:

```rust
// note: different variable name intended here for illustration purposes
non_whitespace_found = non_whitespace_found || !is_whitespace;
```

But it expextedly fails tests. Maybe instead explicitly naming the condition:

```rust
let word_found = non_whitespace_previous && is_whitespace;
```

would be clearer here?

File: src/widgets/reflow.rs [hunk 9]
```diff
@@ -72,155 +72,142 @@ where
...
+                if non_whitespace_previous && is_whitespace
+                    || word_width + symbol_width > self.max_line_width
+                        && current_line.is_empty()
+                        && self.trim
+                    || whitespace_width + symbol_width > self.max_line_width
+                        && current_line.is_empty()
+                        && self.trim
+                    || word_width + whitespace_width + symbol_width > self.max_line_width
+                        && current_line.is_empty()
+                        && !self.trim
```
[the remark this answers is not in the document]
Comment (tranzystorekk) [reply src/widgets/reflow.rs:133]: My current idea is to extract each of these clauses into named variables and then add comments there

File: src/widgets/reflow.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (joshka) [reply src/widgets/reflow.rs:111]: the unclear part is the word "previous" - linguistically it often means "the thing immediately before this point", but here it's supposed to mean "there was some whitespace somewhere before this point". I'm not sure that I follow what you're writing about having two variables for this concept, but word_found seems close enough to the inverted is_first_word to work as a direct rename of `non_whitespace_previous` (I think)

[the remark this answers is not in the document]
Comment (tranzystorekk) [reply src/widgets/reflow.rs:111]: My point is that you might be mistaken and in fact the logic intended here is "the symbol immediately before the current one was a non-whitespace" => we found a delimited word and need to add it to pending line.

File: src/widgets/reflow.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (joshka) [reply src/widgets/reflow.rs:138]: Yeah - that's kinda why I advocate for rewriting this - this might be a good first step in that (understanding what's there already).

If you wanted to really 100% nail the control flow of this, adding tests that incrementally hit every branch / repetition in the method would be the approach I'd take. I'd consider doing it starting from the paragraph in so that replacing this entirely would be possible.

[the remark this answers is not in the document]
Comment (tranzystorekk) [reply src/widgets/reflow.rs:138]: > adding tests that incrementally hit every branch / repetition

tbh, I'd save that for later to avoid exploding the scope of this PR if it's okay 😅

File: src/widgets/reflow.rs (hunk 1 above)
Comment (tranzystorekk) [reply src/widgets/reflow.rs:170]: I'm not sure either, most probably something about eagerly popping incorrectly triggers the continue statement below, but it was hard enough to magic a better equivalent of the original code.

File: src/widgets/reflow.rs (hunk 2 above)
Comment (tranzystorekk) [reply src/widgets/reflow.rs:202]: The simplest way I see is just to wrap the inner if statements in a labeled block with a short circuit if there is no pending word or whitespace.

File: src/widgets/reflow.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (joshka) [reply src/widgets/reflow.rs:113]: I was making a point about two related things - nesting and length
- nesting increases the complexity non-linearly (e.g. an if statement = x2 paths, ...)
- length increases the reading difficulty (having to scroll up and down to work out what a variable's state is)

File: src/widgets/reflow.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (joshka) [reply src/widgets/reflow.rs:111]: Ah got it. I think they actually work out to be the same with one exception.
When running through the loop this starts at false, and is set to false when the symbol is whitespace, and true when the symbol is a character. The exception is when the `continue` in the following part is encountered: https://github.com/ratatui-org/ratatui/blob/3f2f2cd6abf67a04809ff314025a462a3c2e2446/src/widgets/reflow.rs#L168-L172

I'm not certain I understand that part fully however.

But in effect if you have (s)pace and (W)ord symbols:
```
sssWWWsss
FFFTTTFFF
```
What I was saying was that this stays false until we see a word character, then stays true until we see another space (except the bit mentioned). Conceptually it's the transition to a space from a non space that matters, so long as you're not treating the first space as a transition.

TL;DR - you probably understand this better than I do here. :D

File: src/widgets/reflow.rs (hunk 2 above)
Comment (joshka) [reply src/widgets/reflow.rs:202]: I'm not opposed to writing the same body of the if statements twice here if the conditions are clearer - this can possibly be a longer single level if / else-if / ... statement and be more easy to read and maintain. If this was at the top level in a function, I'd be less annoyed by it. But it's nested within a loop and so far down I have to repeatedly scroll up to work out which block it's at the end of.

I'd avoid using labeled blocks generally - they're niche enough that they make this an already complex function more difficult to read rather than easier.

**Comment by tranzystorekk** [comment]
Went ahead and applied changes to resolve the most obvious comments

File: src/widgets/reflow.rs (hunk 2 above)
Comment (joshka) [reply src/widgets/reflow.rs]: explained better elsewhere.

File: src/widgets/reflow.rs [hunk 10]
```diff
@@ -95,119 +207,10 @@ where
...
-            }
-            if !current_line.is_empty() {
-                wrapped_lines.push(current_line);
-            }
-            if wrapped_lines.is_empty() {
-                wrapped_lines.push(vec![]);
-            }
-
-            self.wrapped_lines = Some(wrapped_lines.into_iter());
+            self.process_input(line_symbols);
```
[the remark this answers is not in the document]
Comment (tranzystorekk) [reply src/widgets/reflow.rs:213]: If only it wasn't for that pesky `?` there 😛

[the remark this answers is not in the document]
Comment (joshka) [reply src/widgets/reflow.rs:213]: oh yup. Missed that LOL.

**Comment by joshka** [comment]
Thanks for the PR!
